### ISSUE TITLE
NPE in FrameFlusher toString()

### ISSUE THREAD

**Issue by konikb** [eclipse/jetty.project#2683]
Hello,

after reading text on a websocket, we got a strange exception in AbstractWebSocketConnection.toConnectionString. 
The client sends a message to our server. After reading the message, the server wants to log the org.eclipse.jetty.websocket.api.Session by making a toString(). It gets this exception and our server closes the websocket.

Where may this exception come from ?

Thanks for your help

Bruno Konik

```
2018-06-27 09:52:57,272 [qtp1533734842-144] DEBUG TelServicesWebSocketRemoteEndpoint - client cannot be handled properly, closing web socket WebSocketSession[websocket=JettyListenerEventDriver[com.unigone.telserver.gateway.TelServicesWebSocketRemoteEndpoint],behavior=SERVER,connection=WebSocketServerConnection@e004d6f1[ios=IOState@4b7511dd[OPEN,in,out],f=Flusher@1884f04c[queueSize=0,aggregateSize=0,terminated=null],g=Generator[SERVER,validating],p=Parser@1c43fe4f[ExtensionStack,s=START,c=0,len=78,f=TEXT[len=0,fin=true,rsv=...,masked=true]]]<-SocketChannelEndPoint@756aaa73{/127.0.0.1:65324<->/127.0.0.1:8092,OPEN,fill=-,flush=-,to=0/70000}{io=0/0,kio=0,kro=1}->WebSocketServerConnection@e004d6f1[ios=IOState@4b7511dd[OPEN,in,out],f=Flusher@1884f04c[queueSize=0,aggregateSize=0,terminated=null],g=Generator[SERVER,validating],p=Parser@1c43fe4f[ExtensionStack,s=START,c=0,len=78,f=TEXT[len=0,fin=true,rsv=...,masked=true]]],remote=WebSocketRemoteEndpoint@6078cf84[batching=true],incoming=JettyListenerEventDriver[com.unigone.telserver.gateway.TelServicesWebSocketRemoteEndpoint],outgoing=ExtensionStack[queueSize=0,extensions=[],incoming=org.eclipse.jetty.websocket.common.WebSocketSession,outgoing=org.eclipse.jetty.websocket.server.WebSocketServerConnection]]
java.lang.NullPointerException
	at org.eclipse.jetty.websocket.common.io.FrameFlusher.toString(FrameFlusher.java:340)
	at java.util.Formatter$FormatSpecifier.printString(Formatter.java:2886)
	at java.util.Formatter$FormatSpecifier.print(Formatter.java:2763)
	at java.util.Formatter.format(Formatter.java:2520)
	at java.util.Formatter.format(Formatter.java:2455)
	at java.lang.String.format(String.java:2940)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.toConnectionString(AbstractWebSocketConnection.java:649)
	at org.eclipse.jetty.io.AbstractEndPoint.toConnectionString(AbstractEndPoint.java:477)
	at org.eclipse.jetty.io.AbstractEndPoint.toString(AbstractEndPoint.java:446)
	at java.util.Formatter$FormatSpecifier.printString(Formatter.java:2886)
	at java.util.Formatter$FormatSpecifier.print(Formatter.java:2763)
	at java.util.Formatter.format(Formatter.java:2520)
	at java.util.Formatter.format(Formatter.java:2455)
	at java.lang.String.format(String.java:2940)
	at org.eclipse.jetty.io.AbstractConnection.toString(AbstractConnection.java:266)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.eclipse.jetty.websocket.common.WebSocketSession.toString(WebSocketSession.java:728)
	at com.unigone.telserver.gateway.TelServicesWebSocketRemoteEndpoint.toString(TelServicesWebSocketRemoteEndpoint.java:76)
	at com.unigone.telserver.gateway.TelServicesGateway.dropCallService(TelServicesGateway.java:1507)
	at com.unigone.telserver.gateway.TelServicesGateway.handleTelServiceRequest(TelServicesGateway.java:394)
	at com.unigone.telserver.gateway.TelServicesGateway.handleTelServiceRequest(TelServicesGateway.java:337)
	at com.unigone.telserver.gateway.TelServicesWebSocketRemoteEndpoint.onWebSocketText(TelServicesWebSocketRemoteEndpoint.java:178)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextMessage(JettyListenerEventDriver.java:189)
	at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:66)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextFrame(JettyListenerEventDriver.java:158)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:162)
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:476)
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:220)
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220)
	at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:245)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:560)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:391)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by joakime** [comment]
This seems to be a variation of issue #1603

**Comment by joakime** [comment]
The lines of code referenced by OP (`FrameFlusher` in Jetty 9.4.10.v20180503)...

https://github.com/eclipse/jetty.project/blob/daa59876e6f384329b122929e70a80934569428c/jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java#L334-L342

**Comment by joakime** [comment]
This NPE could happen in the FrameFlusher if the aggregate on line 340 is not null, the thread is tested by line 340, preempted, then different thread releases the aggregate, and then the initial thread then attempts to perform `aggregate.position()`

In short, the FrameFlusher.toString() isn't thread safe.

**Comment by konikb** [comment]
Thanks Joakim, this helps!

**Comment by joakime** [comment]
Commit 33c497a8831904f4fc8d46d02cc78a3cb8f7f6df merged into `jetty-9.4.x`.
